whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java

```diff
@@ -0,0 +1,116 @@
+        final SerialDao serialDao
+    ) {
+        this.deltaTransformer = deltaTransformer;
+        this.nrtmVersionInfoRepository = nrtmVersionInfoRepository;
+        this.snapshotInitializer = snapshotInitializer;
+        this.snapshotFileStreamer = snapshotFileStreamer;
+        this.snapshotFileRepository = snapshotFileRepository;
+        this.snapshotSynchronizer = snapshotSynchronizer;
+        this.nrtmFileUtil = nrtmFileUtil;
+        this.serialDao = serialDao;
+    }
+
+    // TODO: Add a global lock to ensure that no other instance can run until this method exits
+    public Optional<PublishableSnapshotFile> createSnapshot(final NrtmSource source) {
+
+        // Get last version from database.
+        final Optional<NrtmVersionInfo> lastVersion = nrtmVersionInfoRepository.findLastVersion(source);
+        NrtmVersionInfo version;
+        if (lastVersion.isEmpty()) {
+            version = snapshotInitializer.init(source);
+        } else {
+            version = lastVersion.get();
+            if (version.getType() == NrtmDocumentType.DELTA) {
+                version = nrtmVersionInfoRepository.copyAsSnapshotVersion(version);
+            } else {
+                LOGGER.info("Not generating snapshot file since no deltas have been published since v{} with serialID {}",
+                    version.getVersion(), version.getLastSerialId());
+                return Optional.empty();
+            }
+        }
+        // TODO: apply pending deltas
+        //       * if snapshot version > 1...
+        //           - find list of deltas since the last snapshot
+        //           - process them with SnapshotSynchronizer to bring snapshot_objects up to date
+        if (version.getVersion() > 1) {
+            final NrtmVersionInfo lastSnapshot = nrtmVersionInfoRepository.findLastSnapshotVersion(source);
+            final List<SerialEntry> whoisChanges = serialDao.getSerialEntriesBetween(lastSnapshot.getLastSerialId(), version.getLastSerialId());
+            final List<DeltaChange> deltas = deltaTransformer.toDeltaChange(whoisChanges);
+            if (deltas.size() < 1) {
+                LOGGER.error("Changes found but no delta has been published since serialID {}. Now at {}",
+                    lastSnapshot.getLastSerialId(), version.getLastSerialId());
+                return Optional.empty();
+            }
+            snapshotSynchronizer.synchronizeDeltasToSnapshot(deltas, version.getId());
+        }
+        final PublishableSnapshotFile snapshotFile = new PublishableSnapshotFile(version);
+        final ByteArrayOutputStream bos = new ByteArrayOutputStream(4096);
+        try {
+            final String fileName = nrtmFileUtil.fileName(snapshotFile);
+            snapshotFileStreamer.writeSnapshotAsJson(snapshotFile, bos);
+            final String sha256hex = nrtmFileUtil.hashString(bos.toString(StandardCharsets.UTF_8));
+            snapshotFileRepository.save(
+                snapshotFile.getVersionId(),
```

Comment:
Why not pass the snapShotFile mean Directly to save function? Instead of creating new instance of snapShotFile in Save method and here as well?

Reply:
so the snapshot gets published as soon as it can and this thread/transaction can close sooner (which is a good thing). it might take a while to write the file so the transaction will be kept open also (and potentially rolled back). also, other servers won't pick up this change until the read job executes -- so doing it this way means all servers create the file at pretty much the same time. none of this is really necessary, though. it just "feels" a bit neater. thoughts?

Reply:
update: changed this now to generate the file. reason: it takes fewer resources to calculate the hash of a file rather than doing it in memory. it takes a bit longer, though. we'll need to see how long when it's working on a prod, or prod-like machine.

Reply:
hang on -- i don't think that was your question, was it? i thought earlier that i'd need the id for the update notification, so i'd need a new SnapshotFile with the id. doing that a different way now so changed the code -- 'save' returns void.